algebra/src/curves/mod.rs

```diff
@@ -59,320 +28,239 @@ pub mod tests;
-    + for<'a> Add<&'a Self, Output = Self>
-    + for<'a> Sub<&'a Self, Output = Self>
-    + for<'a> AddAssign<&'a Self>
-    + for<'a> SubAssign<&'a Self>
-{
-    type ScalarField: PrimeField + SquareRootField + Into<<Self::ScalarField as PrimeField>::BigInt>;
-    type BaseField: Field;
-    type Affine: AffineCurve<Projective = Self, ScalarField = Self::ScalarField>;
 
-    /// Returns the additive identity.
-    #[must_use]
-    fn zero() -> Self;
-
-    /// Returns a fixed generator of unknown exponent.
-    #[must_use]
-    fn prime_subgroup_generator() -> Self;
-
-    /// Determines if this point is the point at infinity.
-    #[must_use]
-    fn is_zero(&self) -> bool;
-
-    /// Checks that the current point is on curve and is in the
-    /// prime order subgroup
-    #[must_use]
-    fn group_membership_test(&self) -> bool;
+    /// Construct `self` points from a batch of their affine representation.
+    #[inline]
+    fn batch_from_affine<'a>(vec_affine: &'a [Self::AffineRep]) -> Vec<Self> {
+        vec_affine
+            .iter()
+            .map(|&affine| affine.into())
+            .collect::<Vec<_>>()
+    }
 
-    /// Normalizes a slice of projective elements so that
-    /// conversion to affine is cheap.
-    fn batch_normalization(v: &mut [Self]);
+    /// Add an affine point `other` to `self`, using mixed addition formulas.
+    fn add_affine(&self, other: &Self::AffineRep) -> Self;
 
-    fn batch_normalization_into_affine(mut v: Vec<Self>) -> Vec<Self::Affine> {
-        Self::batch_normalization(v.as_mut_slice());
-        v.into_iter().map(|p| p.into_affine()).collect()
-    }
+    /// Add assign an affine point `other` to `self`, using mixed addition formulas.
+    fn add_assign_affine(&mut self, other: &Self::AffineRep);
 
-    /// Checks if the point is already "normalized" so that
-    /// cheap affine conversion is possible.
-    #[must_use]
-    fn is_normalized(&self) -> bool;
+    /// Add, pairwise, the elements of each vector in `to_add`
+    fn add_in_place_affine_many(to_add: &mut [Vec<Self::AffineRep>]);
 
-    /// Doubles this element.
-    #[must_use]
-    fn double(&self) -> Self {
-        let mut copy = *self;
-        copy.double_in_place();
-        copy
+    /// Multiply `self` by the scalar represented by `bits`. 
+    fn mul_bits<S: AsRef<[u64]>>(&self, bits: BitIterator<S>) -> Self {
+        if self.is_zero() {
+            *self
+        } else {
+            Self::mul_bits_affine(&self.into_affine().unwrap(), bits)
+        }
     }
 
-    fn double_in_place(&mut self) -> &mut Self;
+    /// Multiply an affine point `affine` by the scalar represented by `bits`. 
+    fn mul_bits_affine<'a, S: AsRef<[u64]>>(
+        affine: &'a Self::AffineRep,
+        bits: BitIterator<S>,
+    ) -> Self;
 
-    /// Adds an affine element to this element.
-    fn add_assign_mixed(&mut self, other: &Self::Affine);
+    /// Multiply `self` by the cofactor.
+    fn scale_by_cofactor(&self) -> Self;
```

Comment:
We can't as the COFACTOR and COFACTOR_INV params are part of the ModelParameters trait, that we can't access from this function.
If you think it can be important, we can introduce a type Parameters: ModelParameters into the ProjectiveCurve trait.
Afterwards, we would be able to provide default impls for these functions.